games/monop/monop.6
.\"	$OpenBSD: monop.6,v 1.17 2022/02/18 23:17:14 jsg Exp $
.\"
.\" Copyright (c) 1980 The Regents of the University of California.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)monop.6	6.5 (Berkeley) 3/25/93
.\"
.Dd $Mdocdate: February 18 2022 $
.Dt MONOP 6
.Os
.Sh NAME
.Nm monop
.Nd Monopoly game
.Sh SYNOPSIS
.Nm monop
.Op Ar file
.Sh DESCRIPTION
.Nm
is reminiscent of the Parker Brother's game Monopoly, and
monitors a game between 2 to 9 users.
It is assumed that the rules of Monopoly are known.
The game follows the standard rules, with the exception that,
if a property goes up for auction and there are only two solvent players,
no auction is held and the property remains unowned.
.Pp
The game, in effect, lends the player money,
so it is possible to buy something which you cannot afford.
However, as soon as players go into debt,
they must
.Dq fix the problem ,
i.e., make themselves solvent, before play can continue.
If this is not possible, the player's property reverts to their debtee,
either a player or the bank.
A player can resign at any time to any person or the bank,
which puts the property back on the board, unowned.
.Pp
Any time that the response to a question is a
.Em string ,
e.g., a name, place or person, you can type
.Sq Ic \&?
to get a list of valid answers.
It is not possible to input a negative number, nor is it ever necessary.
.Pp
.Em A Summary of Commands :
.Bl -tag -width item
.It Ic quit
quit game.
It asks you if you're sure.
.It Ic print
print current board.
The columns have the following meanings (column headings are the same for the
.Ic where ,
.Ic own holdings ,
and
.Ic holdings
commands):
.Pp
.Bl -tag -width indent -compact
.It Name
The first ten characters of the name of the square.
.It Own
The
.Em number
of the owner of the property.
.It Price
The cost of the property (if any).
.It Mg
This field has a
.Sq *
in it if the property is mortgaged.
.It \&#
If the property is a Utility or Railroad, this is the number
of such owned by the owner.
If the property is land, this is the number of houses on it.
.It Rent
Current rent on the property.
If it is not owned, there is no rent.
.El
.It Ic where
tell where players are.
A `*' indicates the current player.
.It Ic own holdings
list your own holdings,
i.e., money, get-out-of-jail-free cards, and property.
.It Ic holdings
show holdings list: look at anyone's holdings.
It will ask you whose holdings you wish to look at.
When you are finished, type
.Sq Ic done .
.It Ic mortgage
mortgage property:
sets up a list of mortgageable property, and asks which you wish to mortgage.
.It Ic unmortgage
unmortgage property:
unmortgage mortgaged property.
.It Ic buy
buy houses:
sets up a list of monopolies on which you can buy houses.
If there is more than one, it asks you which you want to buy for.
It then asks you how many for each piece of property,
giving the current amount in parentheses after the property name.
If you build in an unbalanced manner
(a disparity of more than one house within the same monopoly),
it asks you to re-input.
.It Ic sell
sell houses:
sets up a list of monopolies from which you can sell houses.
It operates in an analogous manner to
.Ic buy .
.It Ic card
use a get-out-of-jail-free card to get out of jail.
If you're not in jail, or you don't have one, it tells you so.
.It Ic pay
pay $50 to get out of jail, from whence you are put on Just Visiting.
Difficult to do if you're not there.
.It Ic trade
trade with another player.
It asks you with whom you wish to trade,
and then asks you what each wishes to give up.
You can get a summary at the end and, in all cases,
it asks for confirmation of the trade before doing it.
.It Ic resign
resign to another player or the bank.
If you resign to the bank, all property reverts to its virgin state,
and get-out-of-jail-free cards revert to the deck.
.It Ic save
save the current game in a file for later play.
You can continue play after saving,
either by adding the file in which you saved the game after the
.Nm
command, or by using the
.Ic restore
command (see below).
It will ask you which file you wish to save it in,
and, if the file exists, confirm that you wish to overwrite it.
.It Ic restore
read in a previously saved game from a file.
It leaves the file intact.
.It Ic roll
roll the dice and move forward to your new location.
If you simply hit the <RETURN> key instead of a command,
it is the same as typing
.Ic roll .
.El
.Sh FILES
.Bl -tag -width /usr/share/games/cards.pck -compact
.It Pa /usr/share/games/cards.pck
Chance and Community Chest cards
.El
.Sh AUTHORS
.An Ken Arnold
.Sh BUGS
No command can be given an argument instead of a response to a query.
.Pp
Additional cash cannot be generated during an auction by selling buildings
or by mortgaging property.
.Pp
The trading of mortgaged properties is not billed correctly.
.Pp
Resigning to the bank should trigger the immediate auction of all
possessions.


debian/changelog
linux (6.12.10-76061203.202412060638) oracular; urgency=low

  [ Mainline Build ]
  Mainline build at commit: v6.12.3

 -- Michael Aaron Murphy <michael@mmurphy.dev>  Mon, 20 Jan 2025 09:42:56 +0100


games/monop/monop.c
/*	$OpenBSD: monop.c,v 1.16 2016/01/08 18:20:33 mestre Exp $	*/
/*	$NetBSD: monop.c,v 1.3 1995/03/23 08:34:52 cgd Exp $	*/

/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "monop.def"

static void	getplayers(void);
static void	init_players(void);
static void	init_monops(void);

/*
 *	This program implements a monopoly game
 */
int
main(int ac, char *av[])
{
	num_luck = sizeof lucky_mes / sizeof (char *);

	if (pledge("stdio rpath wpath cpath", NULL) == -1)
		err(1, "pledge");

	init_decks();
	init_monops();
	if (ac > 1) {
		if (!rest_f(av[1]))
			restore();
	}
	else {
		getplayers();
		init_players();
	}
	for (;;) {
		printf("\n%s (%d) (cash $%d) on %s\n", cur_p->name, player + 1,
			cur_p->money, board[(int)cur_p->loc].name);
		printturn();
		force_morg();
		execute(getinp("-- Command: ", comlist));
	}
}

/*
 *	This routine gets the names of the players
 */
static void
getplayers(void)
{
	int	i, j;
	char	buf[257];

blew_it:
	for (;;) {
		if ((num_play = get_int("How many players? ")) <= 1 ||
		    num_play > MAX_PL)
			printf("Sorry. Number must range from 2 to %d\n",
			    MAX_PL);
		else
			break;
	}
	if ((cur_p = play = calloc(num_play, sizeof (PLAY))) == NULL)
		err(1, NULL);
	for (i = 0; i < num_play; i++) {
		do {
			printf("Player %d's name: ", i + 1);
			fgets(buf, sizeof(buf), stdin);
			if ((feof(stdin))) {
				printf("user closed input stream, quitting...\n");
				exit(0);
			}
			buf[strcspn(buf, "\n")] = '\0';
		} while (strlen(buf) == 0);
		if ((name_list[i] = play[i].name = strdup(buf)) == NULL)
			err(1, NULL);
		play[i].money = 1500;
	}
	name_list[i++] = "done";
	name_list[i] = 0;
	for (i = 0; i < num_play; i++)
		for (j = i + 1; j <= num_play; j++)
			if (strcasecmp(name_list[i], name_list[j]) == 0) {
				if (j != num_play)
					printf("Hey!!! Some of those are IDENTICAL!!  Let's try that again...\n");
				else
					printf("\"done\" is a reserved word.  Please try again\n");
				for (i = 0; i < num_play; i++)
					free(play[i].name);
				free(play);
				goto blew_it;
			}
}
/*
 *	This routine figures out who goes first
 */
static void
init_players(void)
{
	int	i, rl, cur_max;
	bool	over = 0;
	int	max_pl = 0;

again:
	putchar('\n');
	for (cur_max = i = 0; i < num_play; i++) {
		printf("%s (%d) rolls %d\n", play[i].name, i+1, rl=roll(2, 6));
		if (rl > cur_max) {
			over = FALSE;
			cur_max = rl;
			max_pl = i;
		}
		else if (rl == cur_max)
			over++;
	}
	if (over) {
		printf("%d people rolled the same thing, so we'll try again\n",
		    over + 1);
		goto again;
	}
	player = max_pl;
	cur_p = &play[max_pl];
	printf("%s (%d) goes first\n", cur_p->name, max_pl + 1);
}
/*
 *	This routine initializes the monopoly structures.
 */
static void
init_monops(void)
{
	MON	*mp;
	int	i;

	for (mp = mon; mp < &mon[N_MON]; mp++) {
		mp->name = mp->not_m;
		for (i = 0; i < mp->num_in; i++)
			mp->sq[i] = &board[(int)mp->sqnums[i]];
	}
}
